Initial submission of TimeSeriesInsights public preview swagger spec and examples

sandshadow opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [X ] The title of the PR is clear and informative.
- [X] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ X] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ X] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ X] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ X] My spec meets the review criteria:
  - [ X] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [X] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [X ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR. 
================
Validation:
1) I have run the swagger-linter plugin in vsCode and the open-api validation tool on the spec.
2) The spec has also passed review by the ARM team and their validation tools.
3) I have used AutoRest to generate a .NET client from the spec and tested all operations with real resources.

Notes on Examples
==========================
I have added examples for all operations for each of our resource types. I have not hooked up the EventSource samples as I am getting validation errors that I believe indicate the tool isn't handling inheritance correctly. If I hook the EventSourcesGet.json example, the tool reports:
{ code: 'RESPONSE_VALIDATION_ERROR',
  message: 'Found errors in validating the response with statusCode "200" for x-ms-example "EventSourcesGet" in operation "EventSources_Get".',
  innerErrors:
   [ { code: 'INVALID_RESPONSE_BODY',
       errors:
        [ { code: 'OBJECT_ADDITIONAL_PROPERTIES',
            params:
             [ [ 'creationTime',
                 'provisioningState',
                 'keyName',
                 'eventHubName',
                 'consumerGroupName',
                 'serviceBusNamespace',
                 'eventSourceResourceId' ] ],
            message: 'Additional properties not allowed: creationTime,provisioningState,keyName,eventHubName,consumerGroupName,serviceBusNamespace,eventSourceResourceId',
            path: [ 'properties' ],
            description: 'Properties of the event source.' } ],
       message: 'Invalid body: Additional properties not allowed: creationTime,provisioningState,keyName,eventHubName,consumerGroupName,serviceBusNamespace,eventSourceResourceId',
       path: [] } ] }

However, the additional properties are members of EventHubEventSourceCommonProperties, which inherit (allOf) AzureEventSourceProperties, which in turn inherit (allOf) EventSourceCommonProperties.

AutoRest understands this inheritance when I use it to generate the .NET Client.

Commit:
Initial commit of TimeSeriesInsights swagger spec for public preview.

Commit:
Add swagger examples for all Environmnet operations

Commit:
Add examples for all operations for all resource types.
The examples for EventSource are not linked into the spec because of validation failures, which I believe are caused by the tool not handling polymorphic resources correctly.

msftclas commented:
@sandshadow,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

vishrutshah commented on arm-timeseriesinsights/2017-02-28-preview/Swagger/timeseriesinsights.json:
```diff
@@ -0,0 +1,1780 @@
+        "tags": [
+          "Environments"
+        ],
+        "operationId": "Environments_ListByResourceGroup",
```
Is this a pageable operation? 

Refer: https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-pageable

sandshadow replied:
Not yet. We have a task to return the collections paged. It's not that urgent because our quotas are pretty small, so it will be hard to get a response that needs paging.

vishrutshah replied:
cool

vishrutshah commented on arm-timeseriesinsights/2017-02-28-preview/Swagger/timeseriesinsights.json:
```diff
@@ -0,0 +1,1780 @@
+        "tags": [
+          "Environments"
+        ],
+        "operationId": "Environments_ListBySubscription",
```
Is this a pageable operation?

Refer: https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-pageable

sandshadow replied:
Not yet.

vishrutshah replied:
cool

vishrutshah commented on arm-timeseriesinsights/2017-02-28-preview/Swagger/timeseriesinsights.json:
```diff
@@ -0,0 +1,1780 @@
+            }
+          },
+          "201": {
+            "description": "The event source was successfully created.",
```
same comment as other long-running operations?

sandshadow replied:
Event sources are currently created synchronously.

vishrutshah replied:
cool

vishrutshah commented on arm-timeseriesinsights/2017-02-28-preview/Swagger/timeseriesinsights.json:
```diff
@@ -0,0 +1,1780 @@
+          }
+        ],
+        "responses": {
+          "200": {
```
what is difference between 200 and 204 here? seems to be the same from swagger model? When would you expect to send 200 v/s 204?

sandshadow replied:
This is from the ARM contract:
The resource provider can return 200 (OK) or 204 (NoContent) to indicate that the operation completed successfully. A 200 (OK) should be returned if the object exists and was deleted successfully; and a 204 (NoContent) should be used if the resource does not exist and the request is well formed.

vishrutshah replied:
Good to know. Thanks!

vishrutshah commented on arm-timeseriesinsights/2017-02-28-preview/Swagger/timeseriesinsights.json:
```diff
@@ -0,0 +1,1780 @@
+          }
+        ],
+        "responses": {
+          "200": {
```
when would you expect to send 204 v/s 200 ?

sandshadow replied:
204 means attempt to delete a reference data set that does not exist. 200 means the reference data set was successfully deleted.

vishrutshah commented on arm-timeseriesinsights/2017-02-28-preview/Swagger/timeseriesinsights.json:
```diff
@@ -0,0 +1,1780 @@
+          "200": {
+            "description": "Reference data sets returned successfully.",
+            "schema": {
+              "$ref": "#/definitions/ReferenceDataSetListResponse"
```
Is this pageable operation?

sandshadow replied:
Not yet.

sandshadow replied:
And we only allow 2 reference data sets per environment, so it will probably never need to be pageable.

vishrutshah replied:
cool

vishrutshah commented on arm-timeseriesinsights/2017-02-28-preview/Swagger/timeseriesinsights.json:
```diff
@@ -0,0 +1,1780 @@
+        "name": {
+          "type": "string",
+          "description": "The name of this SKU.",
+          "enum": [
```
Do we have a default here as it's not required?

sandshadow replied:
It actually should be required. Line 999 (in the original review) marks both name and capacity as required.

vishrutshah commented on arm-timeseriesinsights/2017-02-28-preview/Swagger/timeseriesinsights.json:
```diff
@@ -0,0 +1,1780 @@
+            }
+          },
+          "201": {
+            "description": "The environment create request was accepted. Environment provisioning is an asynchroneous operation. You can periodically get your environment definition and monitor progress via the provisioningState property.",
```
Description says "create request was accepted" but http status code `201` means created not accepted

sandshadow replied:
This follows the ARM spec for creating resources. We follow the same pattern if an update must be performed asynchronously.
Creating or Updating Resources (PUT/PATCH)
The API flow should be to: 
1.	Respond to the initial PUT request with a 201 Created or 200 OK (per normal guidance);
2.	Since provisioning is not complete, the PUT response body MUST contain a provisioningState set to a non-terminal value (e.g. “Accepted”, or “Created”);
3.	Optional: The response headers may include a Azure-AsyncOperation header pointing to an Operation resource (as described below);
4.	Future GETs on the resource that was created should continue to return a 200 Status Code and provisioningState field that is *non-terminal* as long as the provisioning is in progress;
5.	After the provisioning completes, the provisioningState field should transition to one of the terminal states (as described below). 
6.	The provisioningState field should be returned on all future GETs, even after it is complete, until some other operation (e.g. a DELETE or UPDATE) causes it to transition to a non-terminal state.

vishrutshah reviewed (changes requested):
- [ ] Make sure to mark long running operation correctly if any
- [ ] Make sure to mark pageable operation correctly if any
- [ ] Make sure to add `Default` response in each of the operation
- [ ] Make sure to have coorect response status codes for all delete operations. Why 200 & 204 both exists?
- [ ] Look for build job  https://travis-ci.org/Azure/azure-rest-api-specs/jobs/223384270 to resolve linter issues

Thanks!

sandshadow commented on arm-timeseriesinsights/2017-02-28-preview/Swagger/timeseriesinsights.json:
```diff
@@ -0,0 +1,1780 @@
+              "$ref": "#/definitions/EnvironmentResource"
+            }
+          },
+          "201": {
```
Done. Also added for PATCH on Environment, since it may be performed asynchronously as well.

sandshadow commented on arm-timeseriesinsights/2017-02-28-preview/Swagger/timeseriesinsights.json:
```diff
@@ -0,0 +1,1780 @@
+            "description": "Parameters for creating an environment resource."
+          }
+        ],
+        "responses": {
```
This means I need to add a 409 for requests with invalid bodies, right? Do I need to do anything else to mark it as "default"?

sandshadow replied:
Sorry - I mean 400. But your pointer to negative responses in a later comment described how to define the default response.

sandshadow commented on arm-timeseriesinsights/2017-02-28-preview/Swagger/timeseriesinsights.json:
```diff
@@ -0,0 +1,1780 @@
+              "$ref": "#/definitions/EnvironmentResource"
+            }
+          },
+          "404": {
```
Thanks. I have added defaults to all operations. Generally 404, except for PUTs which are 400s.

Commit:
Address PR feedback

vishrutshah commented:
@sandshadow Could you please also update x-ms-examples/spec for the response codes to resolve following error categories:

- [ ] RESPONSE_STATUS_CODE_NOT_IN_EXAMPLE
- [ ] RESPONSE_STATUS_CODE_NOT_IN_SPEC

Reference CI: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/223796765#L1078

vishrutshah reviewed (changes requested):
@sandshadow Thanks for adding `default` response codes and updating respective long runnig operations. Could you please also update x-ms-examples/spec for the response codes to resolve following error categories:

- [ ] RESPONSE_STATUS_CODE_NOT_IN_EXAMPLE
- [ ] RESPONSE_STATUS_CODE_NOT_IN_SPEC

Reference CI: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/223796765#L1078

vishrutshah commented:
@sandshadow Regarding the issue you are facing on polymorphic response validation, could you please let me know which version of [oav](https://github.com/Azure/oav) are you using? Also please shoot an email with your example file to me so that we can look into it? Thanks

Commit:
Remove responses from examples that matched the default response in the spec.
The only error now when running oav validate-example is EventSources_Get. I need to solve the problems with polymorphism and x-ms-client-flatten to fix this.

vishrutshah commented:
@sandshadow Thanks for sending the traces. I'll look into it. Meanwhile if you have fixed the above two errors `RESPONSE_STATUS_CODE_NOT_IN_EXAMPLE` & `RESPONSE_STATUS_CODE_NOT_IN_SPEC`. Let me know we should be good to proceed. Thanks!

Commit:
Remove example for EventSources_Get to prevent errors when running "oav validate-example" until we solve the problems with polymorphic responses.

sandshadow commented:
@vishrutshah I have just added the commit that removes the example for EventSources_Get to prevent errors when running "oav validate-example" until we solve the problems with polymorphic responses. The  RESPONSE_STATUS_CODE_NOT_IN_EXAMPLE & RESPONSE_STATUS_CODE_NOT_IN_SPEC errors should be fixed now.

vishrutshah reviewed (changes requested):
Overall LGTM!

Just few questions: If your service has implemented `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.TimeSeriesInsights/operations` api, please include it here. It's an RPC Violation to not define it. Please be aware that following are the linter error being reported from your service:

- [ ] [OperationsAPIImplementationValidation](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/226238975#L1443)
- [ ] [ServiceDefinitionParameters](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/226238975#L1457)
- [ ] [SecurityDefinitionsStructureValidation](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/226238975#L1471)

sandshadow commented:
@vishrutshah I have not implemented an operations API yet. 
Re:  ServiceDefinitionParameters - I do define a ApiVersionParameter in the global parameters section:
```json
    "ApiVersionParameter": {
      "name": "api-version",
      "in": "query",
      "required": true,
      "type": "string",
      "description": "Version of the API to be used with the client request. Current version is 2017-02-28-preview."
    }, 
```
Re:  SecurityDefinitionsStructureValidation. I do have a Security Definitions object defined.
```json  
"securityDefinitions": {
    "azure_auth": {
      "type": "oauth2",
      "authorizationUrl": "https://login.microsoftonline.com/common/oauth2/authorize",
      "flow": "implicit",
      "description": "Azure Active Directory OAuth2 Flow",
      "scopes": {
        "user_impersonation": "impersonate your user account"
      }
    }
  },
```
Any idea why the linter is complaining about either of these? The message says "A swagger must have security definitions and must adhere to the specific structure." Is there some reference I can look at to see the "specific structure"?

Are any of these linter errors blocking for completing the PR? 

Thanks,
Eric

vishrutshah commented:
@sandshadow Ahaa things looks good but not sure why linter is complaining.

I've put more information onto 
bug https://github.com/Azure/autorest/issues/2129 - SecurityDefinitionsStructureValidation
bug https://github.com/Azure/autorest/issues/2128 - ServiceDefinitionParameters

AutorestCI commented:
No modification for Ruby

AutorestCI commented:
No modification for NodeJS

AutorestCI commented:
No modification for Python
